crates/proto/src/rr/rdata/cert.rs

```diff
@@ -0,0 +1,705 @@
+    /// 17    SM2 signing with SM3 hashing [SM2SM3]              Y       [RFC-cuiling-dnsop-sm2-alg-15][informational]
+    SM2SM3,
+
+    /// 23    GOST R 34.10-2012 [ECC-GOST12]                     Y       [RFC9558][informational]
+    ECCGOST12,
+
+    ///   18-22, 24-122    Unassigned
+    Unassigned(u8),
+}
+
+impl From<u8> for Algorithm {
+    fn from(algorithm: u8) -> Self {
+        match algorithm {
+            0 => Algorithm::Reserved(0),
+            1 => Algorithm::RSAMD5,
+            2 => Algorithm::DH,
+            3 => Algorithm::DSA,
+            4 => Algorithm::ECC,
+            5 => Algorithm::RSASHA1,
+            6 => Algorithm::DSANSEC3SHA1,
+            7 => Algorithm::RSASHA1NSEC3SHA1,
+            8 => Algorithm::RSASHA256,
+            9 => Algorithm::Reserved(9),
+            10 => Algorithm::RSASHA512,
+            11 => Algorithm::Reserved(11),
+            12 => Algorithm::ECCGOST,
+            13 => Algorithm::ECDSAP256SHA256,
+            14 => Algorithm::ECDSAP384SHA384,
+            15 => Algorithm::ED25519,
+            16 => Algorithm::ED448,
+            17 => Algorithm::SM2SM3,
+            18..=22 => Algorithm::Unassigned(algorithm),
+            23 => Algorithm::ECCGOST12,
+            24..=122 => Algorithm::Unassigned(algorithm),
+            252 => Algorithm::INDIRECT,
+            253 => Algorithm::PRIVATEDNS,
+            254 => Algorithm::PRIVATEOID,
+            _ => Algorithm::Unassigned(algorithm),
+        }
+    }
+}
+
+impl From<Algorithm> for u8 {
+    fn from(algorithm: Algorithm) -> Self {
+        match algorithm {
+            Algorithm::Reserved(value) if value == 0 => value,
+            Algorithm::RSAMD5 => 1,
+            Algorithm::DH => 2,
+            Algorithm::DSA => 3,
+            Algorithm::ECC => 4,
+            Algorithm::RSASHA1 => 5,
+            Algorithm::DSANSEC3SHA1 => 6,
+            Algorithm::RSASHA1NSEC3SHA1 => 7,
+            Algorithm::RSASHA256 => 8,
+            Algorithm::Reserved(value) if value == 9 => value,
+            Algorithm::RSASHA512 => 10,
+            Algorithm::Reserved(value) if value == 11 => value,
+            Algorithm::ECCGOST => 12,
+            Algorithm::ECDSAP256SHA256 => 13,
+            Algorithm::ECDSAP384SHA384 => 14,
+            Algorithm::ED25519 => 15,
+            Algorithm::ED448 => 16,
+            Algorithm::SM2SM3 => 17,
+            Algorithm::Unassigned(value) if (18..=22).contains(&value) => value,
+            Algorithm::ECCGOST12 => 23,
+            Algorithm::Unassigned(value) if (24..=122).contains(&value) => value,
+            Algorithm::INDIRECT => 252,
+            Algorithm::PRIVATEDNS => 253,
+            Algorithm::PRIVATEOID => 254,
+            Algorithm::Unassigned(value) => value,
+            Algorithm::Reserved(value) => value,
+        }
+    }
+}
+
+impl fmt::Display for Algorithm {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl CERT {
```

Comment:
Keep the `impl CERT` next to `struct CERT` please, above the consistuent type definitions and their impls.